waku/node/peer_manager/peer_manager.nim

```diff
@@ -357,7 +364,10 @@ proc onPeerMetadata(pm: PeerManager, peerId: PeerId) {.async.} =
       reason = "different clusterId reported: " & $pm.wakuMetadata.clusterId & " vs " & $clusterId
       break guardClauses
 
-    if not metadata.shards.anyIt(pm.wakuMetadata.shards.contains(it)):
+    if (
+      pm.peerHasProtocol(peerId, WakuRelayCodec) and
```

Comment:
This is great but I would move this check to line 347, so that we skip the peer earlier and the node doesn't spend effort dialing a node that will be skipped later on.

Sth like that:
```
  if not pm.peerHasProtocol(peerId, WakuRelayCodec):
    debug "skipping peer because it doesn't have relay mounted", peer_id = peerId
    return
```

That would also help to mitigate any possible protocol issue. f.e., we came across a protobuf incompatibility issue that made the shards number not properly parsed.

Reply:
I think the order is intended. The check in question is two-fold:
* The peer being WakuRelay (you could do that earlier, certainly)
* At least one of the peer shards needs to be in the server's metadata. For this one, you need to wait, at least, until 355 is executed, as that's where you retrieve the metadata from the server.

In any case, if what I said is true, you could skip two checks (moving the last one two places up).

Any opinions?

Reply:
> Any opinions?

The idea of my comment was just to make it a bit simpler and return as early as possible. I believe the `metadata` protocol only applies to "relay" connections.

Btw, feel free to merge the PR if you consider it appropriate. I don't want to be a blocker :D